### ISSUE TITLE
[Feature] NSA optimization roadmap

### ISSUE THREAD

**Issue by hlu1** [sgl-project/sglang#11989]
### Attention algorithm

<img width="2536" height="588" alt="Image" src="https://github.com/user-attachments/assets/b25c1894-2551-4eb0-94c8-307c06d4cb67" />

[Link to original table](https://docs.google.com/spreadsheets/d/1u_AaRCEp2xWO3lxBkXJOgzFm4PG9P-88U6M-Dfa-svI/edit?usp=sharing)

The parts highlighted in blue is work that has been done or in progress.

To summarize:

- Use MHA for short context lengths, as suggested in https://github.com/deepseek-ai/DeepSeek-V3.2-Exp/blob/main/DeepSeek_V3_2.pdf
- Below context lengths 2K, sparse attention is the same as non-sparse attention and we can skip the logits computation and directly generates the indices for sparse MLA kernel or use MHA when possible
- The current flashmla_decode kernel is not well optimized on B200. So a separate dequant kernel + flashmla_sparse_bf16 works better for prefill + fp8 kvcache if the kv cache is not too long compared to the q sequence length. The heuristics will need to be updated after new optimizations to either the prefill or decode kernels, making it a bit hard to use in practice. Detailed analysis is [here](https://docs.google.com/spreadsheets/d/1u_AaRCEp2xWO3lxBkXJOgzFm4PG9P-88U6M-Dfa-svI/edit?gid=1552909022#gid=1552909022)

### Kernel optimizations
- FP8 per tensor sparse MLA kernel from trtllm
- More optimizations to flashmla_decode on B200
- Adaptive MHA attention pathway #11892 https://github.com/sgl-project/sglang/pull/12788 https://github.com/sgl-project/sglang/pull/12964 
- Layernorm optimization #12044
- quantize_k_cache_fast (curr: 3 kernel, 6us). No need to cat, but pass in two separate tensors.
- Optimize `torch.cat([q_nope, q_rope])` by either writing a fast triton/cuda kernel or using torch.compile. It's used for prefill/decode but the prefill one is much bigger and has more room for optimizations. The trtllm kernel supports separate `q_nope` and `q_rope`, but flashmla doesn't. #12215 https://github.com/sgl-project/sglang/pull/13022
- DeepGeMM fp8_mqa_logits optimizations https://github.com/sgl-project/sglang/pull/13402

### Indexer optimizations
- [Decode] Optimize dual stream in Indexer https://github.com/sgl-project/sglang/pull/13546
- [Decode] Move deep_gemm.get_paged_mqa_logits_metadata to init time, similar to attention kernel metadata compute
- [Prefill] Optimize _get_topk_ragged where there are a lot of small kernels. Try multi-stream, torch.compile, and add new kernels when necessary.
- [MTP] Enable nextn = 2/4 in deep_gemm.fp8_paged_mqa_logits, which is faster than the current implementation which uses the kernel with nextn = 1 regardless of mtp size.

### Min latency optimizations
- Enable TP in Attention

**Comment by hlu1** [comment]
Relevant PRs so far: https://github.com/sgl-project/sglang/pull/11892, https://github.com/sgl-project/sglang/pull/11655

**Comment by Johnsonms** [comment]
Add NSA Indexer Testsuite https://github.com/sgl-project/sglang/issues/12509

**Comment by thqq479** [comment]
<img width="1511" height="762" alt="Image" src="https://github.com/user-attachments/assets/bbb46394-1ffe-4445-a872-5509ade8a8e1" />

**Comment by thqq479** [comment]
> H200 uses bf16 kvcache by default. Maybe you could try fp8 kvcache and see if it helps. B200 uses fp8 kvcache by default.

I tried it, but it still doesn't work. Could you please share your DeepGemm test results?

https://github.com/deepseek-ai/DeepGEMM/blob/v2.1.1.post3/tests/test_attention.py

Testing shows that processing long contexts indeed takes a lot of time. So, can v3.2 still offer advantages over 3.1?


Testing FP8 MQA Logits:
 > S=2048, SKV=  4096, H= 64, D=128, CP=1:  960 TFLOPS,   72 us,  484 GB/s | clean:  11 us, 1467 GB/s
 > S=2048, SKV=  4096, H= 64, D=128, CP=0: 1004 TFLOPS,  103 us,  419 GB/s | clean:   7 us, 1180 GB/s
 > S=2048, SKV=  8192, H= 64, D=128, CP=1: 1029 TFLOPS,  134 us,  389 GB/s | clean:  18 us, 1884 GB/s
 > S=2048, SKV=  8192, H= 64, D=128, CP=0: 1081 TFLOPS,  222 us,  347 GB/s | clean:   7 us, 1273 GB/s
 > S=2048, SKV= 16384, H= 64, D=128, CP=1: 1088 TFLOPS,  253 us,  343 GB/s | clean:  28 us, 2400 GB/s
 > S=2048, SKV= 16384, H= 64, D=128, CP=0: 1126 TFLOPS,  458 us,  317 GB/s | clean:   7 us, 1256 GB/s
 > S=2048, SKV= 32768, H= 64, D=128, CP=1: 1130 TFLOPS,  486 us,  320 GB/s | clean:  47 us, 2858 GB/s
 > S=2048, SKV= 32768, H= 64, D=128, CP=0: 1144 TFLOPS,  931 us,  302 GB/s | clean:   7 us, 1211 GB/s
 > S=2048, SKV= 65536, H= 64, D=128, CP=1: 1145 TFLOPS,  960 us,  307 GB/s | clean:  90 us, 2981 GB/s
 > S=2048, SKV= 65536, H= 64, D=128, CP=0: 1147 TFLOPS, 1888 us,  294 GB/s | clean:   7 us, 1191 GB/s
 > S=2048, SKV=131072, H= 64, D=128, CP=1: 1150 TFLOPS, 1912 us,  299 GB/s | clean: 174 us, 3087 GB/s
 > S=2048, SKV=131072, H= 64, D=128, CP=0: 1100 TFLOPS, 3967 us,  277 GB/s | clean:   8 us, 1076 GB/s
 > S=4096, SKV=  4096, H= 64, D=128, CP=1:  990 TFLOPS,  139 us,  495 GB/s | clean:  19 us, 1727 GB/s
 > S=4096, SKV=  4096, H= 64, D=128, CP=0:  990 TFLOPS,  139 us,  495 GB/s | clean:  19 us, 1745 GB/s
 > S=4096, SKV=  8192, H= 64, D=128, CP=1: 1061 TFLOPS,  259 us,  397 GB/s | clean:  35 us, 1926 GB/s
 > S=4096, SKV=  8192, H= 64, D=128, CP=0: 1098 TFLOPS,  375 us,  363 GB/s | clean:  18 us, 1823 GB/s
 > S=4096, SKV= 16384, H= 64, D=128, CP=1: 1112 TFLOPS,  495 us,  346 GB/s | clean:  51 us, 2638 GB/s
 > S=4096, SKV= 16384, H= 64, D=128, CP=0: 1128 TFLOPS,  853 us,  319 GB/s | clean:  18 us, 1859 GB/s
 > S=4096, SKV= 32768, H= 64, D=128, CP=1: 1135 TFLOPS,  969 us,  317 GB/s | clean:  92 us, 2913 GB/s
 > S=4096, SKV= 32768, H= 64, D=128, CP=0: 1141 TFLOPS, 1807 us,  300 GB/s | clean:  18 us, 1849 GB/s
 > S=4096, SKV= 65536, H= 64, D=128, CP=1: 1142 TFLOPS, 1926 us,  301 GB/s | clean: 176 us, 3046 GB/s
 > S=4096, SKV= 65536, H= 64, D=128, CP=0: 1114 TFLOPS, 3826 us,  283 GB/s | clean:  20 us, 1722 GB/s
 > S=4096, SKV=131072, H= 64, D=128, CP=1: 1080 TFLOPS, 4071 us,  277 GB/s | clean: 346 us, 3104 GB/s
 > S=4096, SKV=131072, H= 64, D=128, CP=0:  992 TFLOPS, 8730 us,  248 GB/s | clean:  21 us, 1579 GB/s

Testing FP8 Paged MQA Logits:
 > BSZ= 64, NextN=1, H=64, D=128, L=  8192:  310 TFLOPS,  28 us, 2592 GB/s | clean:  12 us,  177 GB/s
 > BSZ= 64, NextN=1, H=64, D=128, L= 32768:  425 TFLOPS,  79 us, 3540 GB/s | clean:  10 us,  787 GB/s
 > BSZ= 64, NextN=2, H=64, D=128, L=  8192:  580 TFLOPS,  29 us, 2517 GB/s | clean:  14 us,  285 GB/s
 > BSZ= 64, NextN=2, H=64, D=128, L= 32768:  797 TFLOPS,  85 us, 3418 GB/s | clean:  13 us, 1301 GB/s
 > BSZ=128, NextN=1, H=64, D=128, L=  8192:  367 TFLOPS,  46 us, 3074 GB/s | clean:  14 us,  283 GB/s
 > BSZ=128, NextN=1, H=64, D=128, L= 32768:  460 TFLOPS, 146 us, 3827 GB/s | clean:  12 us, 1327 GB/s

**Comment by yuan-luo** [comment]
https://github.com/sgl-project/sglang/pull/19148

**Comment by thqq479** [comment]
> ### Attention algorithm
> <img alt="Image" width="2000" height="464" src="https://private-user-images.githubusercontent.com/14827759/504473213-b25c1894-2551-4eb0-94c8-307c06d4cb67.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODcxNDg2NDIsIm5iZiI6MTc4NzE0ODM0MiwicGF0aCI6Ii8xNDgyNzc1OS81MDQ0NzMyMTMtYjI1YzE4OTQtMjU1MS00ZWIwLTk0YzgtMzA3YzA2ZDRjYjY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA4MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwODE5VDE0MDU0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAwNDI2MGZmYTlhMDkxZjFjMTZlMTFlYzc0YTQ0YjAxOWMwM2RiNDE3NDhkNDAzZjY1YzgwNTg4MmI0ODgzOTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.qYB25g0rcqrm0CDbFFt4sCGkVw5bpW1TNTWP60TMEb0">
> [Link to original table](https://docs.google.com/spreadsheets/d/1u_AaRCEp2xWO3lxBkXJOgzFm4PG9P-88U6M-Dfa-svI/edit?usp=sharing)
> 
> The parts highlighted in blue is work that has been done or in progress.
> 
> To summarize:
> 
> * Use MHA for short context lengths, as suggested in https://github.com/deepseek-ai/DeepSeek-V3.2-Exp/blob/main/DeepSeek_V3_2.pdf
> * Below context lengths 2K, sparse attention is the same as non-sparse attention and we can skip the logits computation and directly generates the indices for sparse MLA kernel or use MHA when possible
> * The current flashmla_decode kernel is not well optimized on B200. So a separate dequant kernel + flashmla_sparse_bf16 works better for prefill + fp8 kvcache if the kv cache is not too long compared to the q sequence length. The heuristics will need to be updated after new optimizations to either the prefill or decode kernels, making it a bit hard to use in practice. Detailed analysis is [here](https://docs.google.com/spreadsheets/d/1u_AaRCEp2xWO3lxBkXJOgzFm4PG9P-88U6M-Dfa-svI/edit?gid=1552909022#gid=1552909022)
> 
> ### Kernel optimizations
> * FP8 per tensor sparse MLA kernel from trtllm
> * More optimizations to flashmla_decode on B200
> * Adaptive MHA attention pathway [DeepSeek-V3.2: Add Adaptive MHA Attention Pathway for Short-Sequence Prefill #11892](https://github.com/sgl-project/sglang/pull/11892) [[DeepSeek-V3.2][NSA] Enable MHA Pathway for Short Sequence Prefill on B200 (SM100) #12788](https://github.com/sgl-project/sglang/pull/12788) [[DeepseekV3.2] Deepseek fp8 support for MHA path #12964](https://github.com/sgl-project/sglang/pull/12964)
> * Layernorm optimization [Enable mixed type LayerNorm kernel for NSA indexer #12044](https://github.com/sgl-project/sglang/pull/12044)
> * quantize_k_cache_fast (curr: 3 kernel, 6us). No need to cat, but pass in two separate tensors.
> * Optimize `torch.cat([q_nope, q_rope])` by either writing a fast triton/cuda kernel or using torch.compile. It's used for prefill/decode but the prefill one is much bigger and has more room for optimizations. The trtllm kernel supports separate `q_nope` and `q_rope`, but flashmla doesn't. [[DeepseekV32]: use `_concat_mla_absorb_q_general` to replace `torch.cat` #12215](https://github.com/sgl-project/sglang/pull/12215) [[Deepseek V3.2] Use torch.compile to speed up torch.cat in nsa #13022](https://github.com/sgl-project/sglang/pull/13022)
> * DeepGeMM fp8_mqa_logits optimizations [[sgl-kernel] chore: update deepgemm version #13402](https://github.com/sgl-project/sglang/pull/13402)
> 
> ### Indexer optimizations
> * [Decode] Optimize dual stream in Indexer [[Deepseek V3.2] Optimize use of dual_stream in nsa_indexer/attention #13546](https://github.com/sgl-project/sglang/pull/13546)
> * [Decode] Move deep_gemm.get_paged_mqa_logits_metadata to init time, similar to attention kernel metadata compute
> * [Prefill] Optimize _get_topk_ragged where there are a lot of small kernels. Try multi-stream, torch.compile, and add new kernels when necessary.
> * [MTP] Enable nextn = 2/4 in deep_gemm.fp8_paged_mqa_logits, which is faster than the current implementation which uses the kernel with nextn = 1 regardless of mtp size.
> 
> ### Min latency optimizations
> * Enable TP in Attention

I would like to know what the value of S1 is in this chart? And why is there such a critical point? @hlu1
